File: app/src/main/java/com/test/shoppingapp/di/ViewModelModule.java
package com.test.shoppingapp.di;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import com.test.shoppingapp.shoppinglist.ShoppingListViewModel;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
abstract class ViewModelModule {

    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(ShoppingViewModelFactory factory);


    @Binds
    @IntoMap
    @ViewModelKey(ShoppingListViewModel.class)
    abstract ViewModel bindShoppingListViewModel(ShoppingListViewModel homeViewModel);
}

File: app/src/main/java/com/test/shoppingapp/di/ViewModelModule.kt
package com.test.shoppingapp.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import com.test.shoppingapp.shoppinglist.ShoppingListViewModel

import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
internal abstract class ViewModelModule {

    @Binds
    internal abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory


    @Binds
    @IntoMap
    @ViewModelKey(ShoppingListViewModel::class)
    internal abstract fun bindShoppingListViewModel(shoppingListViewModel: ShoppingListViewModel): ViewModel
}
